### PR TITLE
Initial Implementation of Tagliatelle

### PR DESCRIPTION
**Pull request by andyHa**

### CODE REVIEW

File: src/main/java/sirius/tagliatelle/compiler/Parser.java [hunk 1]
```diff
@@ -0,0 +1,623 @@
...
+            reader.consume();
+            skipWhitespaces();
+            whenFalse = parseExpression();
+        }
+
+        return new TenaryOperation(baseExpression, whenTrue, whenFalse);
+    }
+
+    /**
+     * Parses a {@link #conjunction()} and supports arbitrary many disjunctions ('||') or <b>the nooble operator</b>,
```
Comment (sabieber) [inline src/main/java/sirius/tagliatelle/compiler/Parser.java]: nooble -> noodle

File: src/main/java/sirius/tagliatelle/emitter/Emitter.java [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+     *
+     * @return a deep copy of this emitter, which can safely be modified without changing this emitter
+     */
+    @Nonnull
+    public abstract Emitter copy();
+
+    /**
+     * Performs static optimizations at compile time.
+     * <p>
+     * Certain emitters an perform optimizations (evaluate constant expressions, combine constant outputs etc.) to
```
Comment (sabieber) [inline src/main/java/sirius/tagliatelle/emitter/Emitter.java]: an -> can

File: src/main/java/sirius/tagliatelle/expression/OperationAnd.java [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+ */
+
+package sirius.tagliatelle.expression;
+
+import sirius.tagliatelle.rendering.LocalRenderContext;
+
+/**
+ * Represents a boolean <tt>and</tt>.
+ */
+public class OperationAnd extends BooleanOperation {
```
Comment (sabieber) [inline src/main/java/sirius/tagliatelle/expression/OperationAnd.java]: OperationAnd -> AndOperation

File: src/main/java/sirius/tagliatelle/expression/OperationEquals.java [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+import sirius.tagliatelle.rendering.LocalRenderContext;
+
+import java.util.Objects;
+
+/**
+ * Creates a <tt>==</tt> (in terms of {@link Object#equals(Object)}) operation.
+ * <p>
+ * Note that this can also be inverted to <tt>!=</tt>.
+ */
+public class OperationEquals extends Expression {
```
Comment (sabieber) [inline src/main/java/sirius/tagliatelle/expression/OperationEquals.java]: OperationEquals -> EqualsOperation

File: src/main/java/sirius/tagliatelle/expression/OperationOr.java [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+ */
+
+package sirius.tagliatelle.expression;
+
+import sirius.tagliatelle.rendering.LocalRenderContext;
+
+/**
+ * Represents a boolean <tt>or</tt>.
+ */
+public class OperationOr extends BooleanOperation {
```
Comment (sabieber) [inline src/main/java/sirius/tagliatelle/expression/OperationOr.java]: OperationOr -> OrOperation

File: src/main/java/sirius/tagliatelle/macros/MacroI18n.java [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+import sirius.tagliatelle.rendering.LocalRenderContext;
+
+import javax.annotation.Nonnull;
+import java.util.List;
+
+/**
+ * Represents <tt>i18n(String)</tt> which is essentially a call to {@link NLS#get(String)}.
+ */
+@Register
+public class MacroI18n implements Macro {
```
Comment (sabieber) [inline src/main/java/sirius/tagliatelle/macros/MacroI18n.java]: MacroI18n -> I18nMacro

File: src/main/java/sirius/tagliatelle/macros/MacroIsFilled.java [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+import sirius.tagliatelle.rendering.LocalRenderContext;
+
+import javax.annotation.Nonnull;
+import java.util.List;
+
+/**
+ * Represents <tt>isFilled(Object)</tt> which is a call to {@link Strings#isFilled(Object)}.
+ */
+@Register
+public class MacroIsFilled implements Macro {
```
Comment (sabieber) [inline src/main/java/sirius/tagliatelle/macros/MacroIsFilled.java]: MacroIsFilled -> IsFilledMacro

File: src/main/java/sirius/tagliatelle/tags/TagArg.java [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+import sirius.tagliatelle.expression.ConstantNull;
+import sirius.tagliatelle.expression.ConstantString;
+import sirius.tagliatelle.expression.Expression;
+
+import javax.annotation.Nonnull;
+
+/**
+ * Handles <tt>i:arg</tt> which specifies a template argument.
+ */
+public class TagArg extends TagHandler {
```
Comment (sabieber) [inline src/main/java/sirius/tagliatelle/tags/TagArg.java]: TagArg -> ArgTag

File: src/main/java/sirius/tagliatelle/tags/TagBlock.java [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+import sirius.tagliatelle.emitter.CompositeEmitter;
+import sirius.tagliatelle.emitter.ConstantEmitter;
+import sirius.tagliatelle.emitter.Emitter;
+
+import javax.annotation.Nonnull;
+
+/**
+ * Handles <tt>i:block</tt> which specifies a template section passed into a tag invocation.
+ */
+public class TagBlock extends TagHandler {
```
Comment (sabieber) [inline src/main/java/sirius/tagliatelle/tags/TagBlock.java]: TagBlock -> BlockTag

File: src/main/java/sirius/tagliatelle/tags/TagExtensions.java [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+import sirius.tagliatelle.Template;
+import sirius.tagliatelle.emitter.CompositeEmitter;
+import sirius.web.templates.Templates;
+
+import javax.annotation.Nonnull;
+
+/**
+ * Handles <tt>i:extensions</tt> which invokes all extensions with the given name.
+ */
+public class TagExtensions extends TagInvoke {
```
Comment (sabieber) [inline src/main/java/sirius/tagliatelle/tags/TagExtensions.java]: TagExtensions -> ExtensionsTag

File: src/main/java/sirius/tagliatelle/tags/TagFor.java [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+import sirius.tagliatelle.emitter.CompositeEmitter;
+import sirius.tagliatelle.emitter.LoopEmitter;
+import sirius.tagliatelle.expression.Expression;
+
+import javax.annotation.Nonnull;
+
+/**
+ * Handles <tt>i:for</tt> which emits its body for each item in an {@link Iterable}.
+ */
+public class TagFor extends TagHandler {
```
Comment (sabieber) [inline src/main/java/sirius/tagliatelle/tags/TagFor.java]: TagFor -> ForTag

File: src/main/java/sirius/tagliatelle/tags/TagIf.java [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+
+package sirius.tagliatelle.tags;
+
+import sirius.tagliatelle.emitter.CompositeEmitter;
+import sirius.tagliatelle.emitter.ConditionalEmitter;
+
+/**
+ * Handles <tt>i:if</tt> which emits its body if a condition is met.
+ */
+public class TagIf extends TagHandler {
```
Comment (sabieber) [inline src/main/java/sirius/tagliatelle/tags/TagIf.java]: TagIf -> IfTag

File: src/main/java/sirius/tagliatelle/tags/TagInvoke.java [hunk 13]
```diff
@@ -0,0 +1,94 @@
...
+import sirius.tagliatelle.compiler.CompileException;
+import sirius.tagliatelle.emitter.CompositeEmitter;
+import sirius.tagliatelle.expression.Expression;
+
+import javax.annotation.Nonnull;
+
+/**
+ * Handles <tt>i:invoke</tt> which invokes or inlines a given template.
+ */
+public class TagInvoke extends TagHandler {
```
Comment (sabieber) [inline src/main/java/sirius/tagliatelle/tags/TagInvoke.java]: TagInvoke -> InvokeTag

File: src/main/java/sirius/tagliatelle/tags/TagPragma.java [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+
+import sirius.kernel.di.std.Register;
+import sirius.tagliatelle.emitter.CompositeEmitter;
+
+import javax.annotation.Nonnull;
+
+/**
+ * Handles <tt>i:pragma</tt> which defines a pragma (key / value pair) for a template.
+ */
+public class TagPragma extends TagHandler {
```
Comment (sabieber) [inline src/main/java/sirius/tagliatelle/tags/TagPragma.java]: TagPragma -> PragmaTag

File: src/main/java/sirius/tagliatelle/tags/TagRaw.java [hunk 15]
```diff
@@ -0,0 +1,42 @@
...
+import sirius.tagliatelle.emitter.CompositeEmitter;
+import sirius.tagliatelle.emitter.RawEmitter;
+
+import javax.annotation.Nonnull;
+
+/**
+ * Handles <tt>i:raw</tt> which sets the <tt>escaper</tt> to
+ * {@link sirius.tagliatelle.rendering.GlobalRenderContext#escapeRAW(String)} while emitting its body.
+ */
+public class TagRaw extends TagHandler {
```
Comment (sabieber) [inline src/main/java/sirius/tagliatelle/tags/TagRaw.java]: TagRaw -> RawTag

File: src/main/java/sirius/tagliatelle/tags/TagRender.java [hunk 16]
```diff
@@ -0,0 +1,52 @@
...
+import sirius.kernel.di.std.Register;
+import sirius.tagliatelle.emitter.BlockEmitter;
+import sirius.tagliatelle.emitter.CompositeEmitter;
+
+import javax.annotation.Nonnull;
+
+/**
+ * Handles <tt>i:render</tt> which emits the block with the given name.
+ */
+public class TagRender extends TagHandler {
```
Comment (sabieber) [inline src/main/java/sirius/tagliatelle/tags/TagRender.java]: TagRender -> RenderTag

File: src/main/java/sirius/tagliatelle/TemplateArgument.java [hunk 17]
```diff
@@ -0,0 +1,63 @@
...
+ */
+
+package sirius.tagliatelle;
+
+import sirius.tagliatelle.expression.Expression;
+
+import javax.annotation.Nullable;
+
+/**
+ * Represents a argument definition for a template.
```
Comment (meg1502) [inline src/main/java/sirius/tagliatelle/TemplateArgument.java]: typo 'aN argument'

File: src/main/java/sirius/tagliatelle/compiler/CompilationContext.java [hunk 18]
```diff
@@ -0,0 +1,597 @@
...
+            stack.remove(stack.size() - 1);
+        } else {
+            error(position, "Cannot pop from empty stack");
+        }
+    }
+
+    /**
+     * Returns the maximal stack depth required when rendering this template.
+     *
+     * @return the maximal required stack depth to render this template
```
Comment (meg1502) [inline src/main/java/sirius/tagliatelle/compiler/CompilationContext.java]: maximal -> maximum

File: src/main/java/sirius/tagliatelle/compiler/Compiler.java [hunk 19]
```diff
@@ -0,0 +1,549 @@
...
+     */
+    public Compiler(CompilationContext context, String input) {
+        super(new LookaheadReader(new StringReader(input)), context);
+        this.input = input;
+    }
+
+    /**
+     * Compiles the given source into the template in the given {@link CompilationContext}.
+     * <p>
+     * Note that this can only invoked once per instance.
```
Comment (meg1502) [inline src/main/java/sirius/tagliatelle/compiler/Compiler.java]: can only invoked --> can only be invoked

File: src/main/java/sirius/tagliatelle/compiler/CompilationContext.java [hunk 20]
```diff
@@ -0,0 +1,597 @@
...
+    }
+
+    /**
+     * Tries to resolve a variable name into a stack location.
+     *
+     * @param variableName the name of the variable
+     * @return the type and the stack location of the variable or an empty optional, if no such variable exists
+     */
+    public Optional<Tuple<Class<?>, Integer>> findLocal(String variableName) {
+        for (int i = stack.size() - 1; i >= 0; i--) {
```
Comment (andyHa) [inline src/main/java/sirius/tagliatelle/compiler/CompilationContext.java]: How?

File: src/main/java/sirius/tagliatelle/compiler/Compiler.java [hunk 21]
```diff
@@ -0,0 +1,549 @@
...
+            offset++;
+        }
+
+        return isAtText(offset, waitFor);
+    }
+
+    /**
+     * Processes a tag.
+     * <p>
+     * If the tag is built in (i:) or a one in a taglib, an appropriate {@link TagHandler} is created and invoked.
```
Comment (meg1502) [inline src/main/java/sirius/tagliatelle/compiler/Compiler.java]: I dont quite understand the first part of the sentence

File: src/main/java/sirius/tagliatelle/compiler/Parser.java [hunk 22]
```diff
@@ -0,0 +1,623 @@
...
+            } else {
+                break;
+            }
+        }
+
+        return result;
+    }
+
+    /**
+     * Parses an {@link #atom()} which can be folloed by one or more method calls.
```
Comment (meg1502) [inline src/main/java/sirius/tagliatelle/compiler/Parser.java]: typo 'followed'

File: src/main/java/sirius/tagliatelle/compiler/Parser.java [hunk 23]
```diff
@@ -0,0 +1,623 @@
...
+
+        if ("1".equals(value)) {
+            return ConstantInt.ONE;
+        }
+
+        if ("-1".equals(value)) {
+            return ConstantInt.MINUS_ONE;
+        }
+
+        return new ConstantInt(Integer.parseInt(value));
```
Comment (meg1502) [inline src/main/java/sirius/tagliatelle/compiler/Parser.java]: why is '1' and '-1' parsed explicitly before using parseInt?

File: src/main/java/sirius/tagliatelle/compiler/RawHandler.java [hunk 24]
```diff
@@ -0,0 +1,42 @@
...
+ */
+
+package sirius.tagliatelle.compiler;
+
+import sirius.kernel.di.std.Register;
+import sirius.tagliatelle.emitter.Emitter;
+import sirius.tagliatelle.emitter.RawEmitter;
+
+/**
+ * Handles <tt>{@literal @}raw {}</tt>.
```
Comment (meg1502) [inline src/main/java/sirius/tagliatelle/compiler/RawHandler.java]: the javadoc states this handles `@raw {}` but the code looks like the expected syntax is `@raw() {}`

File: src/main/java/sirius/tagliatelle/compiler/Parser.java (hunk 23 above)
Comment (andyHa) [inline src/main/java/sirius/tagliatelle/compiler/Parser.java]: to not create an instance eacht time - these are quite common...

File: src/main/java/sirius/tagliatelle/compiler/RawHandler.java [hunk 25]
```diff
@@ -0,0 +1,42 @@
...
+ * Handles <tt>{@literal @}raw() {}</tt>.
+ *
+ * @see RawEmitter
+ */
+@Register(classes = ExpressionHandler.class)
+public class RawHandler extends ExpressionHandler {
+
+    @Override
+    public boolean shouldProcess(Compiler compiler) {
+        return compiler.isAtText(0, "@raw()");
```
Comment (sabieber) [inline src/main/java/sirius/tagliatelle/compiler/RawHandler.java]: i dont think its a good idea to check for the braces here. this restricts the usage of whitespaces that would be possible according to the process method. The JavaDoc and process is fine and matches the other handlers, just this shouldProcess check should only check for "@raw"

File: src/main/java/sirius/tagliatelle/emitter/InlineTemplateEmitter.java [hunk 26]
```diff
@@ -0,0 +1,82 @@
...
+public class InlineTemplateEmitter extends Emitter {
+
+    private final Template template;
+    protected Emitter body;
+
+    /**
+     * Creates a new instance at the given position, which references the given template and wraps the given body.
+     * @param startOfBlock the position where the inlining took place
+     * @param template the original template where the emitters of <tt>body</tt> were defined
+     * @param body the inner emitters to invoked within the artificial render stack
```
Comment (meg1502) [inline src/main/java/sirius/tagliatelle/emitter/InlineTemplateEmitter.java]: typo

File: src/main/java/sirius/tagliatelle/emitter/InlineTemplateEmitter.java [hunk 27]
```diff
@@ -0,0 +1,82 @@
...
+    @Override
+    public Emitter copy() {
+        return new InlineTemplateEmitter(startOfBlock, template, body.copy());
+    }
+
+    @Override
+    public Emitter reduce() {
+        this.body = body.reduce();
+        // When outputting a constant emitter, nothing can go wrong (TM), therefore we don't need
+        // to maintain the render stack but rather replace this be the constant emitter itself
```
Comment (meg1502) [inline src/main/java/sirius/tagliatelle/emitter/InlineTemplateEmitter.java]: be --> with

File: src/main/java/sirius/tagliatelle/expression/BooleanOperation.java [hunk 28]
```diff
@@ -0,0 +1,63 @@
...
+/**
+ * Represents a binary operation on boolean types.
+ */
+public abstract class BooleanOperation extends Expression {
+
+    protected Expression leftExpression;
+    protected Expression rightExpression;
+
+    /**
+     * Creates a new operation for the given operaands.
```
Comment (meg1502) [inline src/main/java/sirius/tagliatelle/expression/BooleanOperation.java]: typo

File: src/main/java/sirius/tagliatelle/expression/ReadGlobal.java [hunk 29]
```diff
@@ -0,0 +1,68 @@
...
+ *
+ * @see LocalRenderContext#getGlobal
+ */
+public class ReadGlobal extends Expression {
+
+    private int index;
+    private Class<?> type;
+
+    /**
+     * Creates a new instace which reads the given index in the environment list (aka global variable).
```
Comment (meg1502) [inline src/main/java/sirius/tagliatelle/expression/ReadGlobal.java]: instace -> instance

File: src/main/java/sirius/tagliatelle/expression/ReadLocal.java [hunk 30]
```diff
@@ -0,0 +1,71 @@
...
+ *
+ * @see LocalRenderContext#getLocal(int)
+ */
+public class ReadLocal extends Expression {
+
+    private int index;
+    private Class<?> type;
+
+    /**
+     * Creates a new instace which reads the given index in the local stack.
```
Comment (meg1502) [inline src/main/java/sirius/tagliatelle/expression/ReadLocal.java]: instace -> instance

File: src/main/java/sirius/tagliatelle/rendering/RenderException.java [hunk 31]
```diff
@@ -0,0 +1,57 @@
...
+public class RenderException extends Exception {
+
+    private static final long serialVersionUID = -3342628287682148L;
+
+    private RenderException(String message, Exception ex) {
+        super(message, ex);
+    }
+
+    /**
+     * Creates a new error for the given contexnt and root exception.
```
Comment (meg1502) [inline src/main/java/sirius/tagliatelle/rendering/RenderException.java]: contexnt -> context

File: src/main/java/sirius/tagliatelle/rendering/LocalRenderContext.java [hunk 32]
```diff
@@ -0,0 +1,247 @@
...
+            emitter.emit(subContext);
+        }
+
+        return true;
+    }
+
+    /**
+     * Specifies the blocks made available by the caller.
+     *
+     * @param blocks the blocks passed to the tempalte being rendered
```
Comment (yzzo) [inline src/main/java/sirius/tagliatelle/rendering/LocalRenderContext.java]: tempalte -> template

File: src/main/java/sirius/tagliatelle/tags/TagHandler.java [hunk 33]
```diff
@@ -0,0 +1,220 @@
...
+     */
+    public void setTagName(String tagName) {
+        this.tagName = tagName;
+    }
+
+    /**
+     * Invoked once when all attributes are parsed but before the body is compiled.
+     */
+    public void beforeBody() {
+        // NOOP be default
```
Comment (meg1502) [inline src/main/java/sirius/tagliatelle/tags/TagHandler.java]: NOOP by default (?)

File: src/main/java/sirius/tagliatelle/compiler/CompilationContext.java [hunk 34]
```diff
@@ -0,0 +1,597 @@
...
+            temps.add(new PushLocalEmitter(position, temporaryIndex, value));
+            value = new ReadLocal(arg.getType(), temporaryIndex);
+        }
+        return value;
+    }
+
+    /**
+     * Creates a visitor which replaces a reference to the given argument with the given value.
+     *
+     * @param currentIndex the argument (local( index to replace
```
Comment (yzzo) [inline src/main/java/sirius/tagliatelle/compiler/CompilationContext.java]: typo

File: src/main/java/sirius/tagliatelle/compiler/CompilationContext.java [hunk 35]
```diff
@@ -0,0 +1,597 @@
...
+     * @param currentValue the relplacement expression
```
Comment (yzzo) [inline src/main/java/sirius/tagliatelle/compiler/CompilationContext.java]: relplacement -> replacement

File: src/main/java/sirius/tagliatelle/tags/TagHandler.java (hunk 33 above)
Comment (yzzo) [inline src/main/java/sirius/tagliatelle/tags/TagHandler.java]: by

Comment (andyHa) [inline src/main/java/sirius/tagliatelle/tags/TagHandler.java]: only needed in special cases....

**Pull request merged**